code/ui/blocks/src/blocks/ArgTypes.tsx

```diff
@@ -0,0 +1,81 @@
+/* eslint-disable react/destructuring-assignment */
+import type { Parameters, Renderer, StrictArgTypes } from '@storybook/csf';
+import type { ModuleExports } from '@storybook/types';
+import type { FC } from 'react';
+import type { PropDescriptor } from '@storybook/preview-api';
+import { combineParameters, filterArgTypes } from '@storybook/preview-api';
+import type { ArgTypesExtractor } from '@storybook/docs-tools';
+import React from 'react';
+
+import type { SortType } from '../components';
+import { ArgsTable as PureArgsTable, ArgsTableError } from '../components';
+import { useOf } from './useOf';
+
+type ArgTypesParameters = {
+  include?: PropDescriptor;
+  exclude?: PropDescriptor;
+  sort?: SortType;
+};
+
+type ArgTypesProps = ArgTypesParameters & {
+  of: Renderer['component'] | ModuleExports;
+};
+
+// TODO: generalize
+function extractComponentArgTypes(
+  component: Renderer['component'],
+  parameters: Parameters
+): StrictArgTypes {
+  const { extractArgTypes }: { extractArgTypes: ArgTypesExtractor } = parameters.docs || {};
+  if (!extractArgTypes) {
+    throw new Error(ArgsTableError.ARGS_UNSUPPORTED);
+  }
+  return extractArgTypes(component);
+}
+
+function getArgTypesFromResolved(resolved: ReturnType<typeof useOf>, props: ArgTypesProps) {
+  if (resolved.type === 'component') {
+    const {
+      component,
+      projectAnnotations: { parameters },
+    } = resolved;
+    return {
+      argTypes: extractComponentArgTypes(component, parameters),
+      parameters,
+    };
+  }
+
+  if (resolved.type === 'meta') {
+    const {
+      preparedMeta: { component, argTypes, parameters },
+    } = resolved;
+    const componentArgTypes = component && extractComponentArgTypes(component, parameters);
+    const metaArgTypes = filterArgTypes(argTypes, props.include, props.exclude);
+    return {
+      argTypes: combineParameters(componentArgTypes, metaArgTypes) as StrictArgTypes,
```

Comment:
it surprises me that the `argTypes` from `preparedMeta` aren't already a combination of meta and component `argTypes`. shouldn't they be? similar to how `preparedMeta`'s parameters are already a combination of project- and meta annotations.

(I know that I wrote that code so I should know, but I didn't think of that)

Reply:
For some reason I just assumed the `argTypeEnhancers` wouldn't run for meta, but I didn't think hard enough about it or check. You're right, it's probably not necessary 👍

Reply:
You were 100% right.